ipa-core/src/sharding.rs

```diff
@@ -3,9 +3,53 @@ use std::{
     num::TryFromIntError,
 };
 
+use serde::{Deserialize, Serialize};
+
+use crate::helpers::{HelperIdentity, TransportIdentity};
+
+/// This simple trait is used to make aware on what transport dimnsion one is running. Structs like
+/// [`crate::net::client::MpcHelperClient<R>`] use it to know whether they are talking to other
+/// servers as Shards inside a Helper or as a Helper talking to another Helper in a Ring. This
+/// trait can be used to limit the functions exposed by a struct impl depending on the context that
+/// it's being used. Continuing the previous example, the functions a
+/// [`crate::net::client::MpcHelperClient<R>`] provides are dependent on whether it's communicating
+/// with another Shard or another Helper.
+///
+/// This trait is a safety restriction so that structs or traits only expose an API that's
+/// meaningful for their specific context. When used as a generic bound, it also spreads through
+/// the types making it harder to be misused or combining incompatible types, e.g. Using a
+/// [`ShardIndex`] with a [`Ring`].
+pub trait TransportRestriction: Debug + Send + Sync + Clone + 'static {
+    /// The meaningful identity used in this transport dimension.
+    type Identity: TransportIdentity;
+}
+
+/// This marker is used to restrict communication inside a single Helper, with other shards.
+#[derive(Debug, Copy, Clone)]
+pub struct Sharding;
+
+/// This marker is used to restrict communication inter Helpers. This communication usually has
+/// more restrictions. 3 Hosts with the same sharding index are conencted in a Ring.
+#[derive(Debug, Copy, Clone)]
+pub struct Ring;
```

Comment:
I like how `Client<Helper>` and `Client<Shard>` look and I think ring/sharding leaves some room for interpretation and I'd love to be crystal clear here.

Reply:
I'll add more docs

Reply:
This is more about naming rather than documentation

Reply:
How about: `HelperToHelper` and `ShardToShard` ?
Or `Ring` and `ShardToShard` ?

Let me know what you think is clearer.

Reply:
this is def a bikeshedding territory, but this struct names don't matter in isolation because they are never used directly and only used to disambiguate "parent" struct. So `Client<HelperToHelper>` looks less concise than `Client<Helper>` to me